### PR TITLE
[CSDiagnostics] Fix a crasher in MissingGenericArgumentsFailure

### PR DESCRIPTION
**Pull request by theblixguy**
We were crashing when the generic parameter's declaration was missing. Example:

```swift
public func trichotomy<T : Comparable>(x: T, y: T) -> some Comparable {
  if x < y { return -1 } 
  else if x == y { return 0 }
  return 1
}

public func myTri<T: Comparable, U: Comparable> (retval: UnsafeMutablePointer<U>, x: UnsafeMutablePointer<T>, y: UnsafeMutablePointer<T>) {
  retval.initialize(to: trichotomy(x: x, y: y)) // crash
}

print("Hello, World!")
```

Resolves [SR-11236](https://bugs.swift.org/browse/SR-11236).
Resolves: rdar://problem/52774804

### CODE REVIEW

**Comment by theblixguy** [comment]
cc @xedin

**Comment by xedin** [comment]
@theblixguy Better fix here would be to figure out how to retrieve decl context from generic parameter declaration, because it has to have one.

**Comment by theblixguy** [comment]
@xedin Ok let me figure that out, maybe there’s a way to do that via CS

**Comment by xedin** [comment]
@theblixguy I think the simplest fix here would be to integrate your `hasDecl` check into `getDeclContext()` and emit note only if `DC` is not a `nullptr`.

File: validation-test/compiler_crashers_2_fixed/sr11236.swift [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+  if x < y { return -1 } 
+  else if x == y { return 0 }
+  return 1
+}
+
+public func myTri<T: Comparable, U: Comparable> (retval: UnsafeMutablePointer<U>, x: UnsafeMutablePointer<T>, y: UnsafeMutablePointer<T>) {
+  retval.initialize(to: trichotomy(x: x, y: y))
+}
+
+print("Hello, World!")
```
Comment (xedin) [inline validation-test/compiler_crashers_2_fixed/sr11236.swift]: Since the crash is located in the body of `myTri` this `print` is not really necessary.

**Comment by theblixguy** [comment]
@xedin Sounds good, done!

**Review by xedin — APPROVED** [review]

**Comment by xedin** [comment]
Thank you!

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by theblixguy** [comment]
Thanks! i suppose this is also not eligible to cherry pick to 5.1 due to timing?

**Comment by xedin** [comment]
I'm afraid this is indeed too late...

**Comment by theblixguy** [comment]
Ok no problem :)

**Pull request merged**